packages/headless/src/controllers/commerce/sort/core/headless-core-commerce-sort.ts

```diff
@@ -0,0 +1,168 @@
+export {
+  SortBy,
+  SortDirection,
+  buildRelevanceSortCriterion,
+  buildFieldsSortCriterion,
+};
+
+export interface SortProps {
+  /**
+   * The initial state that should be applied to this `Sort` controller.
+   */
+  initialState?: SortInitialState;
+}
+
+export interface CoreSortProps extends SortProps {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  fetchResultsActionCreator: () => AsyncThunkAction<unknown, void, any>;
+}
+
+export interface SortInitialState {
+  /**
+   * The initial sort criterion to register in state.
+   */
+  criterion?: SortCriterion;
+}
+
+function validateSortInitialState(
+  engine: CommerceEngine,
+  state: SortInitialState | undefined
+) {
+  if (!state) {
+    return;
+  }
+
+  const schema = new Schema<SortInitialState>({
+    criterion: sortCriterionDefinition,
+  });
+
+  validateInitialState(engine, schema, state, 'buildSort');
+}
+
+export interface Sort extends Controller {
+  /**
+   * Updates the sort criterion and executes a new query.
+   *
+   * @param criterion - The new sort criterion.
+   */
+  sortBy(criterion: SortCriterion): void;
+
+  /**
+   * Checks whether the specified sort criterion matches the value in state.
+   *
+   * @param criterion - The criterion to compare.
+   * @returns `true` if the passed sort criterion matches the value in state, and `false` otherwise.
+   */
+  isSortedBy(criterion: SortCriterion): boolean;
+
+  /**
+   * Checks whether the specified sort criterion is available.
+   *
+   * @param criterion - The criterion to check for.
+   * @returns `true` if the passed sort criterion is available, and `false` otherwise.
+   */
+  isAvailable(criterion: SortCriterion): boolean;
+
+  /**
+   * A scoped and simplified part of the headless state that is relevant to the `Sort` controller.
+   */
+  state: SortState;
+}
+
+export interface SortState {
+  /**
+   * The current sort criterion.
+   */
+  appliedSort: SortCriterion;
+
+  /**
+   * The available sort criteria.
+   */
+  availableSorts: SortCriterion[];
+}
+
+/**
+ * @internal
+ * Creates a core `Sort` controller instance for commerce solution types.
+ *
+ * @param engine - The headless commerce engine.
+ * @param props - The configurable `Sort` controller properties.
+ * @returns A `Sort` controller instance.
+ */
+export function buildCoreSort(
+  engine: CommerceEngine,
+  props: CoreSortProps
+): Sort {
+  if (!loadSortReducers(engine)) {
+    throw loadReducerError;
+  }
+
+  const {dispatch} = engine;
+  const controller = buildController(engine);
+  const getState = () => engine.state;
+
+  validateSortInitialState(engine, props.initialState);
+
+  const criterion = props.initialState?.criterion;
+
+  if (criterion) {
+    dispatch(applySort(criterion));
+  }
+
+  return {
+    ...controller,
+
+    get state() {
+      return getState().commerceSort;
+    },
+
+    sortBy(criterion: SortCriterion) {
+      dispatch(applySort(criterion));
+      dispatch(updatePage(0));
```

Comment:
What we're doing in headless core is inconsistent IMO, because we handle pagination reset on facet value toggling in the pagination reducer, but we handle pagination reset on sort change in the sort controller.

I personally prefer the former approach, because otherwise the controller is dispatching an action without ensuring that a reducer can handle it, which feels kind of awkward to me.

Reply:
This totally makes sense! Applied your suggestion in https://github.com/coveo/ui-kit/pull/3474/commits/25aa31a11b7ac550ffb2c2eb63cdad7f7bd4ccf1